tests/components/stream/test_hls.py

```diff
@@ -122,3 +124,37 @@ async def test_stream_ended(hass):
 
     # Stop stream, if it hasn't quit already
     stream.stop()
+
+
+async def test_stream_keepalive(hass):
+    """Test hls stream retries the stream when keepalive=True."""
+    await async_setup_component(hass, "stream", {"stream": {}})
+
+    # Setup demo HLS track
+    source = "test_stream_keepalive_source"
+    stream = preload_stream(hass, source)
+    track = stream.add_provider("hls")
+    track.num_segments = 2
+
+    cur_time = 0
+
+    def time_side_effect():
+        nonlocal cur_time
+        if cur_time >= 80:
+            stream.keepalive = False  # Thread should exit and be joinable.
+        cur_time += 40
+        return cur_time
+
+    with patch("av.open") as av_open, patch(
+        "homeassistant.components.stream.worker.time"
+    ) as mock_time:
+        av_open.side_effect = av.error.InvalidDataError(-2, "error")
+        mock_time.time.side_effect = time_side_effect
+        # Request stream
+        request_stream(hass, source, keepalive=True)
+        stream._thread.join()
+        stream._thread = None
+        assert av_open.call_count == 2
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6103615)